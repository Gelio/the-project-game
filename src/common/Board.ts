import { BoardSize } from '../interfaces/BoardSize';

export class Board {
  public readonly size: BoardSize;

  constructor(size: BoardSize) {
    this.size = size;
  }

  public reset() {
    // TODO: implement this
  }
}
