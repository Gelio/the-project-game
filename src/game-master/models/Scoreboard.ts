export class Scoreboard {
  public team1Score: number;
  public team2Score: number;
  public readonly scoreLimit: number;

  constructor(scoreLimit: number) {
    this.scoreLimit = scoreLimit;
    this.team1Score = 0;
    this.team2Score = 0;
  }
}
