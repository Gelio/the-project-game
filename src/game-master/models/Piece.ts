import { Point } from '../../common/Point';

export class Piece {
  public isSham: boolean;

  /**
   * When the piece is picked up, this is the position of the player
   */
  public position: Point;
  public isPickedUp: boolean;
}
