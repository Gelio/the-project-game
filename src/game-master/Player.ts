import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
