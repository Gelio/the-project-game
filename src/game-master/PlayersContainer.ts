import { TeamId } from '../common/TeamId';
import { Player } from './Player';

export class PlayersContainer {
  public readonly players: Player[] = [];

  public addPlayer(player: Player) {
    if (this.players.indexOf(player) !== -1) {
      throw new Error('Player already added');
    }
    this.players.push(player);
  }

  public removePlayer(player: Player) {
    const playerIndex = this.players.indexOf(player);
    if (playerIndex === -1) {
      throw new Error('Player is not added');
    }
    this.players.splice(playerIndex, 1);
  }

  public getPlayersFromTeam(teamId: TeamId) {
    return this.players.filter(player => player.teamId === teamId);
  }

  public getConnectedPlayers() {
    return this.players.filter(player => player.isConnected);
  }

  public getPlayerById(playerId: number) {
    return this.players.find(player => player.playerId === playerId);
  }
}
