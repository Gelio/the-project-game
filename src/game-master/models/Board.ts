import { Point } from '../../common/Point';
import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { Piece } from './Piece';
import { Tile } from './tiles/Tile';

export class Board {
  public readonly size: BoardSize;
  public readonly tiles: Tile[][];
  public readonly pieces: Piece[] = [];

  constructor(size: BoardSize, tiles: Tile[][]) {
    this.size = size;
    this.tiles = tiles;
  }

  public reset() {
    // TODO: implement this
  }

  public getTileAtPosition(position: Point) {
    const tileColumn = this.tiles[position.x];
    if (!tileColumn) {
      throw new Error('Invalid X coordinate');
    }

    const tile = tileColumn[position.y];
    if (!tile) {
      throw new Error('Invalid Y coordinate');
    }

    return tile;
  }

  public addPlayer(player: Player) {
    this.getTileAtPosition(player.position).player = player;
  }

  public removePlayer(point: Point) {
    this.tiles[point.x][point.y].player = null;
  }

  public movePlayer(player: Player, newPosition: Point) {
    const previousTile = this.getTileAtPosition(player.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.player !== player) {
      throw new Error('Old player position corrupted');
    }

    if (newTile.player) {
      throw new Error('Two players cannot stand on the same tile');
    }

    previousTile.player = null;
    newTile.player = player;
  }

  public addPiece(piece: Piece) {
    if (this.pieces.indexOf(piece) !== -1) {
      throw new Error('Piece already added');
    }

    const tile = this.getTileAtPosition(piece.position);
    if (tile.piece) {
      throw new Error('Piece already exists at that position');
    }

    tile.piece = piece;
    this.pieces.push(piece);
  }

  public removePiece(piece: Piece) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    this.pieces.splice(index, 1);

    if (!piece.isPickedUp) {
      const tile = this.getTileAtPosition(piece.position);
      tile.piece = null;
    }
  }

  public movePiece(piece: Piece, newPosition: Point) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    const previousTile = this.getTileAtPosition(piece.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.piece !== piece) {
      throw new Error('Old piece position corrupted');
    }

    if (newTile.piece) {
      throw new Error('Cannot move a piece on a tile which already has one');
    }

    previousTile.piece = null;
    newTile.piece = piece;
    piece.position = newPosition;
  }
}
