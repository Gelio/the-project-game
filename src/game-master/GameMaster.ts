import { createConnection } from 'net';
import { LoggerInstance } from 'winston';

import { bindObjectProperties } from '../common/bindObjectProperties';
import { Communicator } from '../common/Communicator';
import { LoggerFactory } from '../common/logging/LoggerFactory';
import { UITransport } from '../common/logging/UITransport';

import { ActionDelays } from '../interfaces/ActionDelays';
import { BoardSize } from '../interfaces/BoardSize';
import { GameDefinition, TeamSizes } from '../interfaces/GameDefinition';
import { Message } from '../interfaces/Message';
import { Service } from '../interfaces/Service';

import { ActionInvalidMessage } from '../interfaces/messages/ActionInvalidMessage';
import { ActionValidMessage } from '../interfaces/messages/ActionValidMessage';
import {
  GameStartedMessage,
  GameStartedMessagePayload
} from '../interfaces/messages/GameStartedMessage';
import { PlayerAcceptedMessage } from '../interfaces/messages/PlayerAcceptedMessage';
import { PlayerDisconnectedMessage } from '../interfaces/messages/PlayerDisconnectedMessage';
import { PlayerHelloMessage } from '../interfaces/messages/PlayerHelloMessage';
import { PlayerRejectedMessage } from '../interfaces/messages/PlayerRejectedMessage';
import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';
import { RegisterGameRequest } from '../interfaces/requests/RegisterGameRequest';
import { RegisterGameResponse } from '../interfaces/responses/RegisterGameResponse';

import { registerUncaughtExceptionHandler } from '../registerUncaughtExceptionHandler';

import { PeriodicPieceGenerator } from './board-generation/PeriodicPieceGenerator';

import { Game } from './Game';
import { GameMasterState } from './GameMasterState';
import { Player } from './Player';
import { PlayersContainer } from './PlayersContainer';

import { UIController } from './ui/UIController';

export interface GameMasterOptions {
  serverHostname: string;
  serverPort: number;
  gameName: string;
  gameDescription: string;
  gamesLimit: number;
  teamSizes: TeamSizes;
  pointsLimit: number;
  boardSize: BoardSize;
  shamChance: number;
  generatePiecesInterval: number;
  piecesLimit: number;
  resultFileName: string;
  actionDelays: ActionDelays;
  timeout: number;
  registrationTriesLimit: number;
}

export class GameMaster implements Service {
  private readonly options: GameMasterOptions;
  private communicator: Communicator;
  private game: Game;
  private playersContainer: PlayersContainer;
  private state: GameMasterState;

  private readonly uiController: UIController;
  private readonly loggerFactory: LoggerFactory;
  private periodicPieceGenerator: PeriodicPieceGenerator;
  private logger: LoggerInstance;

  private readonly messageHandlers: { [type: string]: Function } = {
    PLAYER_HELLO: this.handlePlayerHelloMessage,
    PLAYER_DISCONNECTED: this.handlePlayerDisconnectedMessage
  };

  constructor(
    options: GameMasterOptions,
    uiController: UIController,
    loggerFactory: LoggerFactory
  ) {
    this.options = options;
    this.uiController = uiController;
    this.loggerFactory = loggerFactory;
    this.state = GameMasterState.Connecting;

    bindObjectProperties(this.messageHandlers, this);
    this.destroy = this.destroy.bind(this);
    this.handleMessage = this.handleMessage.bind(this);
  }

  public init() {
    this.initUI();
    this.initLogger();
    this.initGame();

    const { serverHostname, serverPort } = this.options;

    const socket = createConnection(
      {
        host: serverHostname,
        port: serverPort
      },
      () => {
        this.logger.info(`Connected to the server at ${serverHostname}:${serverPort}`);
      }
    );

    this.communicator = new Communicator(socket, this.logger);
    this.communicator.bindListeners();

    this.communicator.once('close', this.handleServerDisconnection.bind(this));

    this.registerGame();
  }

  public destroy() {
    if (this.game.hasStarted) {
      this.stopGame();
    }

    this.communicator.destroy();
    this.uiController.destroy();
  }

  private handleServerDisconnection() {
    this.logger.warn('Disconnected from the server. Closing...');

    this.destroy();
  }

  private async handleMessage<T>(message: Message<T>) {
    if (this.messageHandlers[message.type] !== undefined) {
      return this.messageHandlers[message.type](message);
    }

    const result = this.game.processMessage(message);
    if (!result.valid) {
      const actionInvalidMessage: ActionInvalidMessage = {
        type: 'ACTION_INVALID',
        recipientId: message.senderId,
        senderId: -1,
        payload: {
          reason: result.reason
        }
      };

      return this.communicator.sendMessage(actionInvalidMessage);
    }

    const actionValidMessage: ActionValidMessage = {
      type: 'ACTION_VALID',
      recipientId: message.senderId,
      senderId: -1,
      payload: {
        delay: result.delay
      }
    };

    this.communicator.sendMessage(actionValidMessage);
    this.communicator.sendMessage(await result.responseMessage);
  }

  private handlePlayerHelloMessage(message: PlayerHelloMessage) {
    this.logger.verbose(`Received player ${message.payload.temporaryId} hello message`);

    try {
      const assignedPlayerId = this.tryAcceptPlayer(message);

      const playerAcceptedMessage: PlayerAcceptedMessage = {
        type: 'PLAYER_ACCEPTED',
        senderId: -1,
        recipientId: message.payload.temporaryId,
        payload: {
          assignedPlayerId
        }
      };

      this.communicator.sendMessage(playerAcceptedMessage);
      this.tryStartGame();
    } catch (e) {
      const error: Error = e;

      this.logger.verbose(`Player ${message.payload.temporaryId} rejected. Reason: ${e.message}`);

      const playerRejectedMessage: PlayerRejectedMessage = {
        type: 'PLAYER_REJECTED',
        senderId: -1,
        recipientId: message.payload.temporaryId,
        payload: {
          reason: error.message
        }
      };

      return this.communicator.sendMessage(playerRejectedMessage);
    }
  }

  private async registerGame() {
    const game: GameDefinition = {
      name: this.options.gameName,
      description: this.options.gameDescription,
      teamSizes: this.options.teamSizes,
      boardSize: this.options.boardSize,
      goalLimit: this.options.pointsLimit,
      delays: this.options.actionDelays
    };
    const registerGameMessage: RegisterGameRequest = {
      type: 'REGISTER_GAME_REQUEST',
      senderId: -1,
      payload: game
    };

    this.communicator.sendMessage(registerGameMessage);

    const listGamesResponse = <RegisterGameResponse>await this.communicator.waitForSpecificMessage(
      (msg: MessageWithRecipient<RegisterGameResponse>) => msg.type === 'REGISTER_GAME_RESPONSE'
    );

    this.handleRegisterGameResponse(listGamesResponse);
  }

  private handleRegisterGameResponse(message: RegisterGameResponse) {
    if (message.payload.registered === true) {
      this.logger.verbose('Received game registered response');
      this.updateState(GameMasterState.WaitingForPlayers);

      this.communicator.on('message', this.handleMessage);

      return;
    }

    if (this.options.registrationTriesLimit === 0) {
      throw new Error('Failed to register new game, limit of tries reached');
    }

    this.options.registrationTriesLimit--;

    this.logger.error(
      `Failed to register new game! Next attempt will be made in 10 seconds. Attempts left: ${
        this.options.registrationTriesLimit
      }`
    );

    setTimeout(() => this.registerGame(), 10000);
  }

  private tryAcceptPlayer(message: PlayerHelloMessage) {
    const teamPlayers = this.game.playersContainer.getPlayersFromTeam(message.payload.teamId);

    if (this.game.hasStarted) {
      const disconnectedPlayer = teamPlayers.find(
        player => !player.isConnected && player.isLeader === message.payload.isLeader
      );

      if (!disconnectedPlayer) {
        throw new Error('Game already started and no more slots free');
      }

      disconnectedPlayer.isConnected = true;

      return disconnectedPlayer.playerId;
    }

    if (teamPlayers.length >= this.options.teamSizes[message.payload.teamId]) {
      throw new Error('Team is full');
    }

    if (message.payload.isLeader && teamPlayers.find(player => player.isLeader)) {
      throw new Error('Team already has a leader');
    }

    if (
      !message.payload.isLeader &&
      teamPlayers.length + 1 === this.options.teamSizes[message.payload.teamId]
    ) {
      throw new Error('Team does not have a leader');
    }

    const newPlayer = new Player();
    newPlayer.playerId = this.game.getNextPlayerId();
    newPlayer.teamId = message.payload.teamId;
    newPlayer.isLeader = message.payload.isLeader;
    newPlayer.isBusy = false;
    newPlayer.isConnected = true;

    this.game.addPlayer(newPlayer);

    return newPlayer.playerId;
  }

  private handlePlayerDisconnectedMessage(message: PlayerDisconnectedMessage) {
    this.logger.verbose('Received player disconnected message');
    const disconnectedPlayer = this.playersContainer.getPlayerById(message.payload.playerId);

    if (!this.game.hasStarted) {
      if (disconnectedPlayer) {
        this.game.removePlayer(disconnectedPlayer);
        this.uiController.updateBoard(this.game.board);
      }

      return;
    }

    if (!disconnectedPlayer) {
      return;
    }

    disconnectedPlayer.isConnected = false;

    const connectedPlayers = this.game.playersContainer.getConnectedPlayers();
    if (connectedPlayers.length === 0) {
      this.logger.info('All players disconnected, disconnecting from the server');
      this.destroy();
      //TODO check if GM should try to start new game
    }
  }

  private initGame() {
    this.playersContainer = new PlayersContainer();

    this.game = new Game(
      this.options.boardSize,
      this.options.pointsLimit,
      this.logger,
      this.uiController,
      this.playersContainer
    );
    this.uiController.updateBoard(this.game.board);

    this.periodicPieceGenerator = new PeriodicPieceGenerator(
      this.game,
      {
        checkInterval: this.options.generatePiecesInterval,
        piecesLimit: this.options.piecesLimit,
        shamChance: this.options.shamChance
      },
      this.logger
    );
  }

  private tryStartGame() {
    const connectedPlayersCount = this.playersContainer.players.length;
    const requiredPlayersCount = this.options.teamSizes['1'] + this.options.teamSizes['2'];

    if (connectedPlayersCount < requiredPlayersCount) {
      return;
    }

    try {
      this.startGame();
    } catch (error) {
      this.logger.error(`Cannot start game - ${error.message}`);
    }
  }

  private startGame() {
    this.logger.info('Game is starting...');

    const team1Players = this.game.playersContainer.getPlayersFromTeam(1);
    const team2Players = this.game.playersContainer.getPlayersFromTeam(2);
    const team1Leader = team1Players.find(player => player.isLeader);
    const team2Leader = team2Players.find(player => player.isLeader);

    if (!team1Leader || !team2Leader) {
      throw new Error('Game cannot start without both leaders');
    }

    this.game.setPlayersPositions();
    this.periodicPieceGenerator.init();
    const gameStartedPayload: GameStartedMessagePayload = {
      teamInfo: {
        1: {
          players: team1Players.map(player => player.playerId),
          leaderId: team1Leader.playerId
        },
        2: {
          players: team2Players.map(player => player.playerId),
          leaderId: team2Leader.playerId
        }
      }
    };
    this.game.start();
    this.updateState(GameMasterState.InGame);

    this.playersContainer.players.forEach(player => {
      const message: GameStartedMessage = {
        senderId: -1,
        recipientId: player.playerId,
        type: 'GAME_STARTED',
        payload: gameStartedPayload
      };

      this.communicator.sendMessage(message);
    });

    this.logger.info('Game started');
  }

  private stopGame() {
    this.periodicPieceGenerator.destroy();
    this.game.stop();
    this.updateState(GameMasterState.Finished);
  }

  private updateState(state: GameMasterState) {
    this.state = state;
    this.uiController.render();
  }

  private initUI() {
    this.uiController.init();
    this.updateState(this.state);
  }

  private initLogger() {
    if (!this.uiController) {
      throw new Error('Cannot create logger without UI controller');
    }

    const uiTransport = new UITransport(this.uiController.log.bind(this.uiController));
    this.logger = this.loggerFactory.createLogger([uiTransport]);
    registerUncaughtExceptionHandler(this.logger);
    this.logger.info('Logger initiated');
  }
}
