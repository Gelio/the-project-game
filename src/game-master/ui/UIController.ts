import * as blessed from 'blessed';

import { Service } from '../../interfaces/Service';

import { config } from '../config';
import { GameMasterOptions } from '../GameMaster';
import { PlayersContainer } from '../PlayersContainer';

import { Board } from '../models/Board';
import { Scoreboard } from '../models/Scoreboard';

import { BoardFormatter } from './BoardFormatter';

export type BoxFactoryFn = (options: blessed.Widgets.BoxOptions) => blessed.Widgets.BoxElement;

export class UIController implements Service {
  private readonly screen: blessed.Widgets.Screen;
  private readonly boardFormatter: BoardFormatter;
  private readonly boxFactoryFn: BoxFactoryFn;

  private boardBox: blessed.Widgets.BoxElement;
  private infoBox: blessed.Widgets.BoxElement;
  private logsBox: blessed.Widgets.BoxElement;

  constructor(screen: blessed.Widgets.Screen, boxFactoryFn: BoxFactoryFn) {
    this.screen = screen;
    this.boardFormatter = new BoardFormatter();
    this.boxFactoryFn = boxFactoryFn;
  }

  public init() {
    this.boardBox = this.boxFactoryFn({
      width: '50%',
      height: '100%',
      border: {
        type: 'line'
      },
      tags: true
    });
    this.boardBox.setContent(`${config.uiLabelStyle}Board{/}`);

    this.infoBox = this.boxFactoryFn({
      left: '50%',
      width: '50%',
      height: '30%',
      border: {
        type: 'line'
      },
      tags: true
    });
    this.infoBox.setContent(`${config.uiLabelStyle}Info{/}`);

    this.logsBox = this.boxFactoryFn({
      left: '50%',
      top: '30%',
      width: '50%',
      height: '70%',
      border: {
        type: 'line'
      },
      tags: true,
      scrollable: true
    });
    this.logsBox.setContent(`${config.uiLabelStyle}Logs{/}`);

    this.screen.append(this.boardBox);
    this.screen.append(this.infoBox);
    this.screen.append(this.logsBox);
    this.render();
  }

  public destroy() {
    this.screen.destroy();
  }

  public render() {
    this.screen.render();
  }

  public log(lines: string | string[]) {
    this.logsBox.pushLine(lines);
    this.logsBox.setScrollPerc(100);
    this.render();
  }

  public updateBoard(board: Board) {
    this.boardBox.setContent(`${config.uiLabelStyle}Board{/}`);
    const tiles = board.tiles;

    for (let y = 0; y < tiles[0].length; y++) {
      const line = tiles.map(tilesRow => this.boardFormatter.displayTile(board.size, tilesRow[y]));
      this.boardBox.pushLine(line.join(''));
    }

    // FIXME: check if `render` is needed and when should it be called
    this.render();
  }

  public updateGameInfo(
    currentRound: number,
    gameMasterOptions: GameMasterOptions,
    board: Board,
    scoreboard: Scoreboard,
    playersContainer: PlayersContainer
  ) {
    // TODO: add tests

    this.infoBox.setContent(`${config.uiLabelStyle}Info{/}`);

    this.infoBox.pushLine(`Round: ${currentRound} / ${gameMasterOptions.gamesLimit}`);
    this.infoBox.pushLine(
      `{blue-fg}Blue{/} team score: ${scoreboard.team1Score} / ${scoreboard.scoreLimit}`
    );
    this.infoBox.pushLine(
      `{red-fg}Red{/} team score: ${scoreboard.team2Score} / ${scoreboard.scoreLimit}`
    );

    const blueTeamPlayers = playersContainer.getPlayersFromTeam(1).length;
    const blueTeamCapacity = gameMasterOptions.teamSizes[1];
    this.infoBox.pushLine('');
    this.infoBox.pushLine(
      `{blue-fg}Blue{/} team players: ${blueTeamPlayers} / ${blueTeamCapacity}`
    );
    const redTeamPlayers = playersContainer.getPlayersFromTeam(2).length;
    const redTeamCapacity = gameMasterOptions.teamSizes[2];
    this.infoBox.pushLine(`{red-fg}Red{/} team players: ${redTeamPlayers} / ${redTeamCapacity}`);

    this.infoBox.pushLine('');
    const pickedUpPiecesCount = board.pieces.filter(piece => piece.isPickedUp).length;
    const shamCount = board.pieces.filter(piece => piece.isSham).length;
    this.infoBox.pushLine(
      `Pieces: ${board.pieces.length} (${pickedUpPiecesCount} picked up, ${shamCount} shams)`
    );
  }
}
