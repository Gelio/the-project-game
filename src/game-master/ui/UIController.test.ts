import * as blessed from 'blessed';

import { UIController } from './UIController';

import { Board } from '../models/Board';
import { GameMasterOptions } from '../GameMaster';
import { Scoreboard } from '../models/Scoreboard';
import { PlayersContainer } from '../PlayersContainer';
import { Player } from '../Player';
import { Piece } from '../models/Piece';

describe('[GM] UIController', () => {
  let boxes: blessed.Widgets.BoxElement[];
  let screen: blessed.Widgets.Screen;
  let uiController: UIController;

  function createMockScreen(): blessed.Widgets.Screen {
    return <any>{
      render: jest.fn(),
      append: jest.fn(box => boxes.push(box)),
      destroy: jest.fn()
    };
  }

  function createMockBox(): blessed.Widgets.BoxElement {
    let content = '';

    return <any>{
      setContent: jest.fn(val => (content = val)),
      getContent: jest.fn(() => content),
      pushLine: jest.fn(line => (content += `\n${line}`)),
      setScrollPerc: jest.fn()
    };
  }

  function findBox(name: string): blessed.Widgets.BoxElement {
    return <blessed.Widgets.BoxElement>boxes.find(box => box.getContent().includes(name));
  }

  beforeEach(() => {
    boxes = [];
    screen = createMockScreen();
    uiController = new UIController(screen, createMockBox);
  });

  describe('init', () => {
    it('should create and append three boxes', () => {
      uiController.init();

      expect(screen.append).toHaveBeenCalledTimes(3);
    });

    it('should render the screen', () => {
      uiController.init();

      expect(screen.render).toHaveBeenCalled();
    });

    it('should append the logs box', () => {
      uiController.init();

      const logsBox = findBox('Logs');

      expect(logsBox).toBeDefined();
    });

    it('should append the board box', () => {
      uiController.init();

      const boardBox = findBox('Board');

      expect(boardBox).toBeDefined();
    });

    it('should append the info box', () => {
      uiController.init();

      const infoBox = findBox('Info');

      expect(infoBox).toBeDefined();
    });
  });

  describe('destroy', () => {
    beforeEach(() => {
      uiController.init();
    });

    it('should destroy the screen', () => {
      uiController.destroy();

      expect(screen.destroy).toHaveBeenCalled();
    });
  });

  describe('render', () => {
    beforeEach(() => {
      uiController.init();
    });

    it('should render the screen', () => {
      uiController.render();

      expect(screen.render).toHaveBeenCalled();
    });
  });

  describe('log', () => {
    beforeEach(() => {
      uiController.init();
    });

    it('should push and scroll in the logs box', () => {
      const logsBox = findBox('Logs');
      spyOn(logsBox, 'pushLine');
      spyOn(logsBox, 'setScrollPerc');

      uiController.log('test');

      expect(logsBox.pushLine).toHaveBeenCalledWith('test');
      expect(logsBox.setScrollPerc).toHaveBeenCalledWith(100);
    });
  });

  describe('updateBoard', () => {
    let board: Board;

    beforeEach(() => {
      board = new Board(
        {
          goalArea: 2,
          taskArea: 3,
          x: 5
        },
        /**
         * NOTE: for snapshot tests to work the board needs to have team areas covered in goals
         * to prevent random goal generation.
         */
        10
      );

      uiController.init();
    });

    it('should update the board box', () => {
      uiController.updateBoard(board);

      const boardBox = findBox('Board');

      expect(boardBox.getContent()).toMatchSnapshot();
    });

    it('should render the screen', () => {
      uiController.updateBoard(board);

      expect(screen.render).toHaveBeenCalled();
    });
  });

  describe('updateGameInfo', () => {
    let currentRound: number;
    let board: Board;
    let gameMasterOptions: GameMasterOptions;
    let scoreboard: Scoreboard;
    let playersContainer: PlayersContainer;
    let infoBox: blessed.Widgets.BoxElement;

    beforeEach(() => {
      currentRound = 1;

      board = new Board(
        {
          goalArea: 2,
          taskArea: 3,
          x: 5
        },
        5
      );

      gameMasterOptions = <any>{
        gamesLimit: 5,
        teamSizes: {
          1: 2,
          2: 3
        }
      };

      scoreboard = new Scoreboard(board.pointsLimit);

      playersContainer = new PlayersContainer();

      uiController.init();
      infoBox = findBox('Info');
    });

    function runUpdateGameInfo() {
      uiController.updateGameInfo(
        currentRound,
        gameMasterOptions,
        board,
        scoreboard,
        playersContainer
      );
    }

    it('should call render on the screen', () => {
      runUpdateGameInfo();

      expect(screen.render).toHaveBeenCalled();
    });

    it('should match snapshot by default', () => {
      runUpdateGameInfo();

      expect(infoBox.getContent()).toMatchSnapshot();
    });

    it('should match snapshot for the next round', () => {
      currentRound = 2;
      runUpdateGameInfo();

      expect(infoBox.getContent()).toMatchSnapshot();
    });

    it('should match snapshot when scores have changed', () => {
      scoreboard.team1Score = 2;
      scoreboard.team2Score = 5;
      runUpdateGameInfo();

      expect(infoBox.getContent()).toMatchSnapshot();
    });

    it('should match snapshot after players joined', () => {
      const player1 = new Player();
      player1.teamId = 1;

      const player2 = new Player();
      player2.teamId = 2;

      playersContainer.addPlayer(player1);
      playersContainer.addPlayer(player2);

      runUpdateGameInfo();

      expect(infoBox.getContent()).toMatchSnapshot();
    });

    it('should match snapshot after a piece is added', () => {
      const piece = new Piece();
      piece.isPickedUp = true;
      piece.isSham = false;

      runUpdateGameInfo();

      expect(infoBox.getContent()).toMatchSnapshot();
    });
  });
});
