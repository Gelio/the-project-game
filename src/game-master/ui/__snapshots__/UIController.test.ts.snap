// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] UIController updateBoard should update the board box 1`] = `
"{black-fg}{white-bg}{bold}Board{/}
{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}
{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}{blue-bg}{white-fg}G{/}
     
     
     
{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}
{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}{red-bg}{white-fg}G{/}"
`;

exports[`[GM] UIController updateGameInfo should match snapshot after a piece is added 1`] = `
"{black-fg}{white-bg}{bold}Info{/}
Round: 1 / 5
{blue-fg}Blue{/} team score: 0 / 5
{red-fg}Red{/} team score: 0 / 5

{blue-fg}Blue{/} team players: 0 / 2
{red-fg}Red{/} team players: 0 / 3

Pieces: 1 (1 picked up, 0 shams)"
`;

exports[`[GM] UIController updateGameInfo should match snapshot after players joined 1`] = `
"{black-fg}{white-bg}{bold}Info{/}
Round: 1 / 5
{blue-fg}Blue{/} team score: 0 / 5
{red-fg}Red{/} team score: 0 / 5

{blue-fg}Blue{/} team players: 1 / 2
{red-fg}Red{/} team players: 1 / 3

Pieces: 0 (0 picked up, 0 shams)"
`;

exports[`[GM] UIController updateGameInfo should match snapshot by default 1`] = `
"{black-fg}{white-bg}{bold}Info{/}
Round: 1 / 5
{blue-fg}Blue{/} team score: 0 / 5
{red-fg}Red{/} team score: 0 / 5

{blue-fg}Blue{/} team players: 0 / 2
{red-fg}Red{/} team players: 0 / 3

Pieces: 0 (0 picked up, 0 shams)"
`;

exports[`[GM] UIController updateGameInfo should match snapshot for the next round 1`] = `
"{black-fg}{white-bg}{bold}Info{/}
Round: 2 / 5
{blue-fg}Blue{/} team score: 0 / 5
{red-fg}Red{/} team score: 0 / 5

{blue-fg}Blue{/} team players: 0 / 2
{red-fg}Red{/} team players: 0 / 3

Pieces: 0 (0 picked up, 0 shams)"
`;

exports[`[GM] UIController updateGameInfo should match snapshot when scores have changed 1`] = `
"{black-fg}{white-bg}{bold}Info{/}
Round: 1 / 5
{blue-fg}Blue{/} team score: 2 / 5
{red-fg}Red{/} team score: 5 / 5

{blue-fg}Blue{/} team players: 0 / 2
{red-fg}Red{/} team players: 0 / 3

Pieces: 0 (0 picked up, 0 shams)"
`;
